pkg/iacinventory/github.go

```diff
@@ -0,0 +1,241 @@
+	}
+	if len(reposWithCode) == 0 {
+		return fmt.Errorf("no repositories contained any code to be analyzed")
+	}
+	g.repos = reposWithCode
+	return nil
+}
+
+func (g *GithubScanner) Run() (*jnode.Node, error) {
+	var err error
+	// if no authentication options have been set,
+	// we default to reading from the filesystem.
+	if g.User == "" && g.OauthToken == "" {
+		g.User, g.OauthToken, err = githubCredsFromFS()
+		if err != nil {
+			return nil, fmt.Errorf("error getting gh credentials from configuration files: %w", err)
+		}
+		if g.User == "" || g.OauthToken == "" {
+			return nil, fmt.Errorf("internal error: no credentials and no error during iacinventory.Run()")
+		}
+	}
+
+	// fetch the repositories associated with the account
+	if err := g.GetRepos(); err != nil {
+		return nil, fmt.Errorf("error fetching github repositories: %w", err)
+	}
+
+	// massage the data into a sensible format for submission to the API.
+	var out []Repo
+	for name, repo := range g.repos {
+		var tfdirs []string
+		for dir := range repo.TerraformDirs {
+			tfdirs = append(tfdirs, dir)
+		}
+		var cfdirs []string
+		for dir := range repo.CloudformationDirs {
+			cfdirs = append(cfdirs, dir)
+		}
+		var cis []string
+		for _, ci := range repo.CI {
+			cis = append(cis, string(ci))
+		}
+		out = append(out, Repo{
+			Name:               name,
+			CIs:                cis,
+			TerraformDirs:      tfdirs,
+			CloudformationDirs: cfdirs,
+		})
+	}
+
+	pretty, err := json.MarshalIndent(out, "", "    ")
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal JSON: %w", err)
+	}
+
+	// print some nice, non-json info about notable repos
+	for _, repo := range out {
+		ciMsg := "and has NO CI configuration."
+		if len(repo.CIs) == 1 {
+			ciMsg = fmt.Sprintf("and is configured with %s CI.", repo.CIs[0]) // TODO: include other CI systems
+		}
+		if len(repo.CIs) > 1 {
+			ciMsg = "and is configured with multiple CI systems."
+		}
+		if len(repo.TerraformDirs) > 0 {
+			log.Infof("[%s]: contains %d Terraform directories %s\n", repo.Name, len(repo.TerraformDirs), ciMsg)
+		}
+		if len(repo.CloudformationDirs) > 0 {
+			log.Infof("[%s]: contains %d CloudFormation directories %s\n", repo.Name, len(repo.CloudformationDirs), ciMsg)
+		}
+	}
+
+	return jnode.FromJSON(pretty)
+}
+
+func (g *GithubScanner) Stop() error {
+	return nil
+}
+
+// githubCredsFromFS reads the `gh` configuration file on disk to get GitHub credentials.
+func githubCredsFromFS() (username string, oauthToken string, retErr error) {
+	var ghConfig map[string]interface{}
+	var ghConfigFile string
+
+	user, err := user.Current()
+	if err != nil {
+		retErr = fmt.Errorf("unable to get OS user: %w", err)
+		return
+	}
+
+	// Github credentials can exist in one of two files: ~/.config/gh/hosts.yml,
+	// or ~/.config/gh/config.yml. Creds _should_ only exist in the latter, but
+	// on some systems it appears in the former for whatever reason.
+	ghConfigFile = filepath.Join(user.HomeDir, ".config/gh/hosts.yml")
```

Comment:
I think https://github.com/mitchellh/go-homedir is preferred, though to be honest I'm not completely sure.